core/trino-spi/src/main/java/io/trino/spi/predicate/SortedRangeSet.java

```diff
@@ -1042,6 +1046,81 @@ else if (thatNextRangeIndex == thatRangeCount) {
         return new SortedRangeSet(type, inclusive, blockBuilder.build(), unionIsDiscreteSet(that, resultRangeIndex > 0));
     }
 
+    SortedRangeSet linearDiscreteSetUnion(SortedRangeSet that)
+    {
+        return linearDiscreteSetUnion(this, that);
+    }
+
+    private SortedRangeSet linearDiscreteSetUnion(SortedRangeSet leftRangeSet, SortedRangeSet rightRangeSet)
+    {
+        int leftRangeCount = leftRangeSet.getRangeCount();
+        int rightRangeCount = rightRangeSet.getRangeCount();
+
+        boolean[] inclusive = new boolean[2 * (leftRangeCount + rightRangeCount)];
+        BlockBuilder blockBuilder = type.createBlockBuilder(null, 2 * (leftRangeCount + rightRangeCount));
+
+        int resultRangeIndex = 0;
+        int leftNextRangeIndex = 0;
+        int rightNextRangeIndex = 0;
+
+        int currentUnionStart = 0;
+        while (leftNextRangeIndex < leftRangeCount && rightNextRangeIndex < rightRangeCount) {
+            int compare = compareValues(
+                    comparisonOperator,
+                    leftRangeSet.sortedRanges,
+                    2 * leftNextRangeIndex,
+                    rightRangeSet.sortedRanges,
+                    2 * rightNextRangeIndex);
+
+            if (compare == 0) {
+                leftNextRangeIndex++;
+                rightNextRangeIndex++;
+            }
+            else if (compare < 0) {
+                leftNextRangeIndex++;
+            }
+            else {
+                int size = leftNextRangeIndex - currentUnionStart;
+                if (size > 0) {
+                    copyBlock(leftRangeSet, currentUnionStart * 2, blockBuilder, inclusive, resultRangeIndex * 2, size);
+                    resultRangeIndex += size;
+                    currentUnionStart = rightNextRangeIndex;
+                }
+
+                // Swap leftRangeSet and rightRangeSet for continue consuming the lower value
+                SortedRangeSet tempSortedSet = leftRangeSet;
+                leftRangeSet = rightRangeSet;
+                rightRangeSet = tempSortedSet;
+
+                int tempNextIndex = leftNextRangeIndex;
+                leftNextRangeIndex = rightNextRangeIndex;
+                rightNextRangeIndex = tempNextIndex;
+
+                int tempRangeCount = leftRangeCount;
+                leftRangeCount = rightRangeCount;
+                rightRangeCount = tempRangeCount;
+            }
+        }
+
+        if (currentUnionStart < leftRangeCount) {
+            int size = leftRangeCount - currentUnionStart;
+            copyBlock(leftRangeSet, currentUnionStart * 2, blockBuilder, inclusive, resultRangeIndex * 2, size);
+            resultRangeIndex += size;
+        }
+
+        if (rightNextRangeIndex < rightRangeCount) {
```

Comment:
nit: could be else if

Reply:
Here can not to be else if. The case `left:  {1,2,3}, right: {10, 11}` will fail. when exists `while`. left, right both may not to be added.